test/common/upstream/original_dst_cluster_test.cc

```diff
@@ -569,6 +569,66 @@ TEST_F(OriginalDstClusterTest, UseHttpHeaderEnabled) {
       2, TestUtility::findCounter(stats_store_, "cluster.name.original_dst_host_invalid")->value());
 }
 
+// Verify original dst cluster can read from HTTP host header. The corner cases are tested in
+// `UseHttpHeaderEnabled`.
+TEST_F(OriginalDstClusterTest, UseHttpAuthorityHeader) {
+  std::string yaml = R"EOF(
+    name: name
+    connect_timeout: 1.250s
+    type: ORIGINAL_DST
+    lb_policy: CLUSTER_PROVIDED
+    original_dst_lb_config:
+      use_http_header: true
+      http_header_name: ":authority"
+  )EOF";
+
+  EXPECT_CALL(initialized_, ready());
+  EXPECT_CALL(*cleanup_timer_, enableTimer(_, _));
+  setupFromYaml(yaml);
+
+  OriginalDstCluster::LoadBalancer lb(cluster_);
+  Event::PostCb post_cb;
+
+  // HTTP header override by `:authority`.
+  TestLoadBalancerContext lb_context1(nullptr, Http::Headers::get().Host.get(), "127.0.0.1:6666");
+  lb_context1.downstream_headers_->setCopy(Http::Headers::get().EnvoyOriginalDstHost,
+                                           "127.0.0.1:5555");
+
+  EXPECT_CALL(membership_updated_, ready());
+  EXPECT_CALL(dispatcher_, post(_)).WillOnce(SaveArg<0>(&post_cb));
+  HostConstSharedPtr host1 = lb.chooseHost(&lb_context1);
+  post_cb();
+  ASSERT_NE(host1, nullptr);
+  EXPECT_EQ("127.0.0.1:6666", host1->address()->asString());
+}
+
+// Verify clearing `use_http_header` is required to consume `http_header_name`.
+TEST_F(OriginalDstClusterTest, DoNotReadFromHttpAuthorityHeaderUnlessSetUseHttpHeader) {
+  std::string yaml = R"EOF(
+    name: name
+    connect_timeout: 1.250s
+    type: ORIGINAL_DST
+    lb_policy: CLUSTER_PROVIDED
+    original_dst_lb_config:
+      # `use_http_header` is not set and default value is false.
+      http_header_name: ":authority"
```

Comment:
emm....thinking of why we didn't raise exception to tell the user, you have invalid config here.

Reply:
Yeah I would fail this case. In a redone API it would be a single message but we can't easily do that so might as well make it more clear to the user.